handout/di/angular2/avoiding_collisions_opaque_token.md

```diff
@@ -0,0 +1,86 @@
+
+`OpaqueToken`s are unique and immutable values which allow developers to avoid collisions of dependency injection token ids.
+
+```
+import { OpaqueToken } from '@angular/core';
+
+const name = 'token';
+const token1 = new OpaqueToken(name);
+const token2 = new OpaqueToken(name);
+
+console.log(token1 === token2); // false
+```
+
+Here, regardless of whether or not the same value is passed to the constructor of the token, it will not generate the same result. Consider this example of two separate modules that are both looking for api configs.
+
+_app/email.module.ts_
+```javascript
+export const apiConfig = 'api-config';
+
+@Injectable()
+export class EmailService {
+  constructor(@Inject(apiConfig) public apiConfig: EmailConfig) { }
+}
+
+@NgModule({
+  providers: [ EmailService ],
+})
+export class EmailModule { }
+```
+
+_app/logger.module.ts_
+```javascript
+
+export const apiConfig = 'api-config';
+
+@Injectable()
+export class LoggerService {
+  constructor(@Inject(apiConfig) public apiConfig: LoggerConfig) { }
+}
+
+@NgModule({
+  providers: [ LoggerService ],
+})
+export class LoggerModule { }
+```
+[View Example](https://plnkr.co/edit/4apXe4XTOy9jDbVIajL3?p=preview)
+
+Both `EmailModule` and `LoggerModule` are attempting to inject their api configuration using the same key. When it comes time to inject the value
+
+_app/app.module.ts_
+```javascript
+@NgModule({
+  ...
+  providers: [
+    { provide: emailApiConfig, useValue: { apiKey: 'email-config', context: 'registration' } },
+    { provide: loggerApiConfig, useValue: { apiKey: 'logger-config' } },
+  ],
+  ...
+})
+export class AppModule { }
+```
+
+Angular overwrites the first `emailApiConfig` value with the `loggerApiConfig` value, since it was provided last. In this case, module implementation details are leaking out to the parent module. Not only that, those details were obfuscated through the module exports and can lead to problematic debugging.
+
+Using `OpaqueToken`, Angular is able to resolve the correct entities.
+
+_app/email.module.ts_
+```javascript
+export const apiConfig = new OpaqueToken('api-config');
+
+@Injectable()
+export class EmailService {
+  constructor(@Inject(apiConfig) public apiConfig: EmailConfig) { }
+}
+
+@NgModule({
+  providers: [
+    EmailService,
+  ],
+})
+export class EmailModule { }
+```
+[View Example](https://plnkr.co/edit/esaYGmJK8vOULxHKCgET?p=preview)
```

Comment:
I've found the code snippets somehow hard to follow, mostly because it's not clear the relationship between `apiConfig` and `emailApiConfig` or `loggerApiConfig`. Also, I think that in a real world scenario, those providers should be defined directly in the feature modules instead of the root module. Take a look at my refactored version of your code [before](https://plnkr.co/edit/vWcZcNAxRFqDkZzL4XEF?p=preview) and [after](https://plnkr.co/edit/56VkwzOVdQsCxRfWVM4Y?p=preview) `OpaqueToken`. What do you think?